library/math/dual.lisp

```diff
@@ -0,0 +1,65 @@
+;;;; dual.lisp
+;;;;
+;;;; dual numbers to enable automatic differentiation
+
+(coalton-library/utils:defstdlib-package #:coalton-library/math/dual
+  (:use
+   #:coalton
+   #:coalton-library/builtin
+   #:coalton-library/classes
+   #:coalton-library/functions
+   #:coalton-library/math/arith)
+  (:export
+   #:Dual
+   #:dual-real
+   #:dual-dual
+   #:sub-duals
+   #:add-duals
+   #:mul-duals))
+
+(in-package #:coalton-library/math/dual)
+
+(coalton-toplevel
+
+ (define-type *Real*
+     (*Real* Integer))
+
+ (define-type *Dual*
+     (*Dual* Integer))
+ 
+ (define-type Dual
+     (Dual *Real* *Dual*))
+ 
+ (declare dual-real (Dual -> Integer))
+ (define (dual-real d1)
+     (match d1
+	    ((Dual (*Real* x) (*Dual* y)) x)))
+ 
+ (declare dual-dual (Dual -> Integer))
+ (define (dual-dual d1)
+     (match d1
+	    ((Dual (*Real* x) (*Dual* y)) y)))
+
+ (declare add-duals (Dual -> Dual -> Dual))
```

Comment:
You ought to implement an instance of the `Num` type class.

```
(define-instance (Num (Dual :t)) 
  (define (+ ...)
    ...)

  (define (- ...)
    ...)

  (define (* ...)
    ...)

  (define (from-int ...)
    ...))
```

Then the normal arithmetic operators will work for dual numbers.